lint-rules.md

```diff
@@ -0,0 +1,1280 @@
+
+**Group**: `completeness`
+
+#### Example
+
+Good:
+
+```wdl
+    meta {
+        description: "Exemplifies the proper grammar for a description string."
+    }
+```
+
+### `nonmatching_output`
+
+The `meta` section should have an `output` key and keys with descriptions for each output of the task/workflow. These must match exactly. i.e. for each named output of a task or workflow, there should be an entry under `meta.output` with that same name. Additionally, these entries should be in the same order (that order is up to the developer to decide). No extraneous output entries are allowed. There should not be any blank lines inside the entire `meta` section.
+
+**Group**: `completeness`
+
+#### Example
+
+Good:
+
+```wdl
+task alignment {
+    meta {
+        description: "Runs the STAR aligner on a set of RNA-Seq FASTQ files"
+        external_help: "https://github.com/alexdobin/STAR/blob/2.7.10a/doc/STARmanual.pdf"
+        output: {
+            star_log: "Summary mapping statistics after mapping job is complete. The statistics are calculated for each read (single- or paired-end) and then summed or averaged over all reads. Note that STAR counts a paired-end read as one read. Most of the information is collected about the UNIQUE mappers. Each splicing is counted in the numbers of splices, which would correspond to summing the counts in SJ.out.tab. The mismatch/indel error rates are calculated on a per base basis, i.e. as total number of mismatches/indels in all unique mappers divided by the total number of mapped bases.",
+            star_bam: "STAR aligned BAM",
+            star_junctions: "File contains high confidence collapsed splice junctions in tab-delimited format. Note that STAR defines the junction start/end as intronic bases, while many other software define them as exonic bases. See `meta.external_help` for file specification.",
+            star_chimeric_junctions: "Tab delimited file containing chimeric reads and associated metadata. See `meta.external_help` for file specification.",
+        }
+    }
+    ...
+    output {
+        File star_log = prefix + ".Log.final.out"
+        File star_bam = prefix + ".Aligned.out.bam"
+        File star_junctions = prefix + ".SJ.out.tab"
+        File? star_chimeric_junctions = prefix + ".Chimeric.out.junction"
+    }
+    ...
+}
+```
+
+### `nonmatching_parameter_meta` && `input_not_sorted`
```

Comment:
- (Strong opinion) I think this lint (`input_not_sorted`) should sort not by WDL type in the `parameter_meta`, but by the actual order they appear in the task declaration. I often group inputs together based on what fits together (e.g. `String sample_name` is next to `File input_vcf` for a task that extracts that sample from the vcf, etc.) so rearranging by type alone would break the scientific cohesion of the user-based groupings

Reply:
I also might've misunderstood this lint, but my understanding was that it would take a task input and sort it based purely on the WDL type in the hierarchy described in the doc. So if my inputs looked like this at first:
```
File vcf
String sample_name

File bam
String read_group

File debug
File interval_list
```
the linted version would look like:
```
File bam
File debug
File interval_list
File vcf
String read_group
String sample_name
```
even though conceptually the "sample_name" and "vcf" inputs should be "near" each other because the "sample_name" references a sample in the vcf. 

In case I misunderstood the lint, here's what the ideal version would look like to me: `nonmatching_parameter_meta` makes sure that the `parameter_meta` input order matches the user-defined on in the task inputs, but would probably turn off `input_not_sorted` because I usually cluster my input based on conceptually similarity, regardless of WDL type/alphabetical order.

Reply:
@adthrasher good points. I also think its reasonable for a user to want an alphabetical sort, and it's probably worth it to include that as an option. Between the 3 mentioned choices (our partially-opinionated-partially-arbitrary typed sort, alphabetical sort, and unsorted), IDK what's best as a default. [tangent] I don't think we've really settled on an exact philosophy for how we determine defaults.